File: src/Algorithm/LeetCode/LeetCode_19_Remove_Nth_Node_From_End_of_List/Solution_Stack.java
package Algorithm.LeetCode.LeetCode_19_Remove_Nth_Node_From_End_of_List;

import java.util.Stack;

class Solution_Stack {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        Stack<ListNode> stack = new Stack<>();

        ListNode dummy = new ListNode(0, head);
        ListNode newHead = dummy;

        while (newHead != null) {
            stack.push(newHead);
            newHead = newHead.next;
        }

        ListNode temp = stack.peek();

        for (int i = 0; i <= n; i++) {
            temp = stack.pop();
        }

        temp.next = temp.next.next;

        return dummy.next;
    }
}

File: src/Algorithm/LeetCode/LeetCode_19_Remove_Nth_Node_From_End_of_List/Solution_Stack.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        stack = []
        dummy_head = ListNode(next=head)
        curr_node = dummy_head

        while curr_node:
            stack.append(curr_node)
            curr_node = curr_node.next

        node = stack[-1]
        for i in range(n + 1):
            node = stack.pop()

        node.next = node.next.next

        return dummy_head.next
